docs/cpu-metrics-in-fleet.md

```diff
@@ -0,0 +1,374 @@
+a counter value representing the total (user-space + kernel-space) duration, in milliseconds, spent by the Agent or Beat
+utilizing the CPU since the process was started. More on how this duration is calculated in the next section.
+
+#### Comparison between metrics in `/proc/$PID/stat` file and in Agent + Beats `/stats` API output
+
+Using the following script for a host running Agent and one child Beat (excluding any monitoring Beats), we can see that
+the metrics in the `/proc/$PID/stat` file match up with those in the Agent + Beats `/stats` API output.
+
+```shell
+#!/bin/bash
+
+BEAT_CPU_MS_TOTAL=$(sudo curl -s -X GET --unix-socket '/opt/Elastic/Agent/data/tmp/PGwsYWcynGUYZEjD872Gs-npqbv-30jS.sock' 'http:/f/stats' | jq '.beat.cpu.total.value')
+AGENT_CPU_MS_TOTAL=$(sudo curl -s http://localhost:6791/stats  | jq '.beat.cpu.total.value')
+
+echo "Stats from API outputs: $(($BEAT_CPU_MS_TOTAL + $AGENT_CPU_MS_TOTAL))";
+
+AGENT_PID=403165
+
+AGENT_USER_TICKS=$(cat /proc/$AGENT_PID/stat | cut -d' ' -f14)
+AGENT_SYSTEM_TICKS=$(cat /proc/$AGENT_PID/stat | cut -d' ' -f15)
+AGENT_TOTAL_TICKS=$(($AGENT_USER_TICKS + $AGENT_SYSTEM_TICKS))
+AGENT_TOTAL_MS=$(($AGENT_TOTAL_TICKS * 1000 / 100))
+
+#echo "Agent total ticks: $AGENT_TOTAL_TICKS"
+#echo "Agent total ms: $AGENT_TOTAL_MS"
+
+BEAT_PID=431834
+
+BEAT_USER_TICKS=$(cat /proc/$BEAT_PID/stat | cut -d' ' -f14)
+BEAT_SYSTEM_TICKS=$(cat /proc/$BEAT_PID/stat | cut -d' ' -f15)
+BEAT_TOTAL_TICKS=$(($BEAT_USER_TICKS + $BEAT_SYSTEM_TICKS))
+BEAT_TOTAL_MS=$(($BEAT_TOTAL_TICKS * 1000 / 100))
+
+#echo "Beat total ticks: $BEAT_TOTAL_TICKS"
+#echo "Beat total ms: $BEAT_TOTAL_MS"
+
+echo "Stats from /proc/PID/stats files: $(($AGENT_TOTAL_MS + $BEAT_TOTAL_MS))"
+```
+
+### Comparison between metrics in Agent + Beats `/stats` API output and in Elasticsearch `metrics-elastic_agent*` indices
+
+This comparison is relatively easy to make.
+
+First, we call the `/stats` APIs on the machine where Agent and its Beats are running. For example, with an Agent running
+one Beat (excluding monitoring Beats):
+
+```shell
+$ sudo curl -s http://localhost:6791/stats  | jq '.beat.cpu.total.value'
+34810
+$ sudo curl -s -X GET --unix-socket '/opt/Elastic/Agent/data/tmp/PGwsYWcynGUYZEjD872Gs-npqbv-30jS.sock' 'http:/f/stats' | jq '.beat.cpu.total.value'
+795000
+```
+
+Then we call the Elasticsearch `_search` API on `metrics-elastic_agent*` indices, keeping the query filters the same as
+the query being done by Fleet UI, but only considering the latest documents for each `elastic_agent.process`, since the
+CPU utilization metrics are counter metrics.
+
+```
+curl -s -u $ES_USER:$ES_PASS -H 'Content-Type: application/json' 'https://test-cpu.es.us-central1.gcp.cloud.es.io:9243/metrics-elastic_agent*/_search' -d '{
+  "sort": [
+    {
+      "@timestamp": {
+        "order": "desc"
+      }
+    }
+  ],
+  "collapse": {
+    "field": "elastic_agent.process"
```

Comment:
Does this account for there being multiple instances of each Beat process? It could depending on if we use the component ID (filestream-default) or the just the process name (e.g. filebeat) as the value.

For reference the default Fleet policy with the system integration and monitoring runs 3 metricbeat instances (system/metrics-default, http/metrics-monitoring, beat/metrics-monitoring) and 2 filebeat instances (filestream-monitoring, log-default).

If this is averaging the 3 metricbeats and 2 filebeats together as if they are one process instead of summing the usage of each distinct process I don't think it is correct.

The memory usage query has this problem as described in https://github.com/elastic/kibana/issues/174458

Reply:
It doesn't.  The value of this field is just `filebeat`.  The query should be using the `component.id` field itself, which has values such as `log-default`.  I will add this fix to the list of suggestions.

Reply:
Looks like this is already tracked as a TODO in https://github.com/elastic/kibana/issues/174458 now that I look closer at it

Reply:
If you're referring to https://github.com/elastic/kibana/issues/174458#issuecomment-1885770207, I added that based on this conversation here :).

Reply:
One more question: did you observe us reporting CPU for the monitoring components?

https://github.com/elastic/kibana/issues/174458#issuecomment-1887717174

There is some evidence we are not reporting them, causing us to undercount CPU and memory usage by omitting the monitoring component contributions.

Reply:
I did not observe us reporting CPU for monitoring components themselves.

Reply:
Added a suggestion for collecting CPU usage metrics for monitoring components.